core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -509,6 +511,13 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorBusy(executorId: String): Unit = synchronized {
     logDebug(s"Clearing idle timer for $executorId because it is now running a task")
     removeTimes.remove(executorId)
+
+    // Executor is added to remove by misjudgment due to async listener making it as idle).
+    // see SPARK-9552
+    if (executorsPendingToRemove.contains(executorId)) {
```

Comment:
Instead of fixing the list of pending executors here, I wonder if it wouldn't be better to change `removeExecutor` to return `false` when the executors asked to be killed were busy? The they wouldn't even be added to this list to start with.

That changes slightly the semantics of the return value, but it also sounds more correct. With your changes, `killExecutors(..., force = false)` will return `true` even if it didn't kill any executors, which sounds wrong.